### PR TITLE
[serve] Improve autoscaling metric collection

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Previously, the number of ongoing requests is recorded every 10 seconds (by default, controlled by `metric_interval_s`) and sent to the controller, then aggregated at the controller to make autoscaling decisions. This can be improved by aggregating at the replica then sending the aggregated results to the controller.
* this allows collecting autoscaling metrics more frequently at each replica actor (e.g. once every 0.5 or 1 second)
* this makes the aggregated results more accurate, since the average is calculated with data collected once every 1s vs 10s, without having to increase the frequency of sending metrics to controller.

Some comparisons (different experiments with different autoscaling parameters)
| Original  | With Changes |
| ------------- | ------------- |
|  Screen Shot 2023-08-07 at 1 06 45 PM | Screen Shot 2023-08-07 at 1 06 12 PM  |
| Screen Shot 2023-08-07 at 1 08 08 PM | Screen Shot 2023-08-07 at 1 07 42 PM  |
| Screen Shot 2023-08-07 at 1 09 21 PM |  Screen Shot 2023-08-07 at 1 09 01 PM |
| Screen Shot 2023-08-07 at 1 11 00 PM | Screen Shot 2023-08-07 at 1 10 35 PM |
| Screen Shot 2023-08-07 at 1 26 04 PM| Screen Shot 2023-08-07 at 1 25 38 PM |
| Screen Shot 2023-08-07 at 6 25 52 PM | Screen Shot 2023-08-07 at 6 25 35 PM | 
|Screen Shot 2023-08-10 at 1 56 36 PM| Screen Shot 2023-08-10 at 1 56 16 PM |
| image| image|

In some cases, there is not a huge difference, but in some cases the new changes obviously stabilize the autoscaling.

## 

Related issue number

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@sihanwang41 @shrekris-anyscale PTAL

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work! I think the approach makes sense at a high level.

File: python/ray/serve/_private/autoscaling_policy.py [hunk 1]
```diff
@@ -43,7 +44,20 @@ def calculate_desired_num_replicas(
     )
 
     # Multiply the distance to 1 by the smoothing ("gain") factor (default=1).
-    smoothed_error_ratio = 1 + ((error_ratio - 1) * autoscaling_config.smoothing_factor)
+    if error_ratio >= 1:
+        smoothing_factor = float(
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/autoscaling_policy.py:48]: Sorry, should have clarified - please ignore the smoothing factor changes! I included them in this PR so I can use the wheel to experiment with both the autoscaling metrics changes and separate upscale + downscale smoothing factors. The changes to review should be outside of `autoscaling_policy.py`.

[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/autoscaling_policy.py:48]: Ah, got it, thanks for clarifying! I'll resolve these threads.

File: python/ray/serve/_private/autoscaling_policy.py [hunk 2]
```diff
@@ -43,7 +44,20 @@ def calculate_desired_num_replicas(
...
+            os.environ.get(
+                "UPSCALE_SMOOTHING_FACTOR", autoscaling_config.smoothing_factor
+            )
+        )
+    else:
+        smoothing_factor = float(
+            os.environ.get(
```
Comment (sihanwang41) [inline python/ray/serve/_private/autoscaling_policy.py:55]: why do we need to define the env variable?

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -533,13 +536,21 @@ def user_health_check():
                 config.metrics_interval_s,
                 process_remote_func,
             )
+            self.metrics_pusher.register_task(
+                lambda: {self.replica_tag: self.get_num_pending_and_running_requests()},
+                RAY_SERVE_AUTOSCALING_METRIC_RECORD_PERIOD_S,
```
Comment (sihanwang41) [inline python/ray/serve/_private/replica.py:541]: nit: RAY_SERVE_REPLICA_AUTOSCALING_METRIC_RECORD_PERIOD_S

File: python/ray/serve/_private/constants.py [hunk 4]
```diff
@@ -228,6 +228,8 @@
 )
 # Serve gauge metric set period.
 RAY_SERVE_GAUGE_METRIC_SET_PERIOD_S = 1
+# Serve autoscaling metric record period.
```
Comment (sihanwang41) [inline python/ray/serve/_private/constants.py:231]: please explicitly mentioning this is for serve replica.

**Comment by zcin** [comment]
Please ignore the smoothing factor changes, the changes to review should be outside of `autoscaling_policy.py`. The smoothing factor changes will be removed from this PR, they are here temporarily for testing.

**Comment by zcin** [comment]
@shrekris-anyscale @sihanwang41 @edoakes What do folks think about this change? In all the autoscaling tests I've ran in the past week, I've run with both the master version and with the changes in this PR, and I think the improvement in stability is pretty consistent. 
The only other side note is I've noticed that this autoscaling generally decides on a lower number of replicas than the existing autoscaling (e.g. the fourth row is a more drastic example, 118 replicas vs 107 replicas), but I think that's because the data is more accurate.

**Review by edoakes — APPROVED** [review]
LGTM

File: python/ray/serve/_private/constants.py [hunk 5]
```diff
@@ -228,6 +228,8 @@
 )
 # Serve gauge metric set period.
 RAY_SERVE_GAUGE_METRIC_SET_PERIOD_S = 1
+# How often autoscaling metrics are recorded on Serve replicas.
+RAY_SERVE_REPLICA_AUTOSCALING_METRIC_RECORD_PERIOD_S = 0.5
```
Comment (edoakes) [inline python/ray/serve/_private/constants.py:232]: What was the previous interval at which we pushed the metrics to the controller? Is this changed? If so, would prefer to make that a separate PR (and test it independently).

**Comment by edoakes** [comment]
Sorry, my comment above didn't register when I submitted review. Pending that comment LGTM to merge.

File: python/ray/serve/_private/constants.py (hunk 5 above)
Comment (zcin) [reply python/ray/serve/_private/constants.py:232]: There are now two separate intervals: one at which we record data for aggregation, and one at which we push the aggregated results to the controller (previously, these were the same). `RAY_SERVE_REPLICA_AUTOSCALING_METRIC_RECORD_PERIOD_S` is now 0.5, but the interval at which we push metrics to the controller hasn't changed - it's default is 10 seconds, and it's tunable by the user.

**Review by shrekris-anyscale — APPROVED** [review]
Nice work, this looks good to me. I only have [one pending nit](https://github.com/ray-project/ray/pull/38107/files#r1286403054). Feel free to merge it or resolve it.

File: python/ray/serve/_private/constants.py (hunk 5 above)
Comment (edoakes) [reply python/ray/serve/_private/constants.py:232]: Got it, that's helpful 👍
